lib/interface-types/src/interpreter/instructions/records.rs

```diff
@@ -0,0 +1,424 @@
+use crate::{
+    ast::{InterfaceType, RecordType, Type, TypeKind},
+    errors::{InstructionError, InstructionErrorKind},
+    interpreter::wasm::values::FlattenInterfaceValueIterator,
+    interpreter::{
+        stack::{Stack, Stackable},
+        wasm::values::InterfaceValue,
+        Instruction,
+    },
+};
+use std::mem::{transmute, MaybeUninit};
+
+/// Build a `InterfaceValue::Record` based on values on the stack.
+///
+/// To fill a record, every field `field_1` to `field_n` must get its
+/// value from the stack with `value_1` to `value_n`. To simplify this
+/// algorithm that also typed-checks values when hydrating, the number
+/// of values to read from the stack isn't known ahead-of-time. Thus,
+/// the `Stack::pop` method cannot be used, and `Stack::pop1` is used
+/// instead. It implies that values are read one after the other from
+/// the stack, in a natural reverse order, from `value_n` to
+/// `value_1`.
+///
+/// Consequently, record fields are filled in reverse order, from
+/// `field_n` to `field_1`.
+///
+/// A basic algorithm would then be:
+///
+/// ```rust,ignore
+/// let mut values = vec![];
+///
+/// // Read fields in reverse-order, from `field_n` to `field_1`.
+/// for field in fields.iter().rev() {
+///     let value = stack.pop1();
+///     // type-check with `field` and `value`, to finally…
+///     values.push(value);
+/// }
+///
+/// InterfaceValue::Record(values.iter().rev().collect())
+/// ```
+///
+/// Note that it is required to reverse the `values` vector at the end
+/// because `InterfaceValue::Record` expects its values to match the
+/// original `fields` order.
+///
+/// Because this approach allocates two vectors for `values`, another
+/// approach has been adopted. `values` is an initialized vector
+/// containing uninitialized values of type
+/// `MaybeUninit<InterfaceValue>`. With this approach, it is possible
+/// to fill `values` from index `n` to `0`. Once `values` is entirely
+/// filled, it is `transmute`d to `Vec<InterfaceType>`.
+///
+/// This latter approach allows to allocate one and final vector to
+/// hold all the record values.
+#[allow(unsafe_code)]
+fn record_lift_(
+    stack: &mut Stack<InterfaceValue>,
+    record_type: &RecordType,
+) -> Result<InterfaceValue, InstructionErrorKind> {
+    let length = record_type.fields.len();
+    let mut values = {
+        // Initialize a vector of length `length` with `MaybeUninit`
+        // values.
+        let mut v = Vec::with_capacity(length);
```

Comment:
I tried, but there is an issue with that. `vec![item; length]` calls `clone` on `item` and I'm not sure it will does what we want.
Also, this code raises an error:

```
    --> lib/interface-types/src/interpreter/instructions/records.rs:61:27
     |
61   |     let mut values = vec![MaybeUninit::<InterfaceValue>::uninit(); length];
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |                           |
     |                           expected an implementor of trait `std::clone::Clone`
     |                           help: consider borrowing here: `&MaybeUninit::<InterfaceValue>::uninit()`
     |
     = note: required because of the requirements on the impl of `std::clone::Clone` for `std::mem::MaybeUninit<interpreter::wasm::values::InterfaceValue>`
```

Adopting the recommandation (`help:` hint) does not help neither.

Also, the code is actually what the macro would more or less generate a similar piece of code. I prefer to understand my code well rather than using magic-shortcut :-D.

If you can come with an alternative, I'm all opened to use it!